### PR TITLE
[Shape Infer] RNN/GRU/LSTM Cell & Sequence - common shape infer and review

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 - RNN/GRU/LSTM Cell & Sequence - common shape infer and review
 - Additional tests
 
 ### Tickets:
 - 97235, 97236
 
#### Shapes of the recurrent operations:

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

Inputs | RNN Cell | GRU Cell | LSTM Cell
-- | -- | -- | --
X | [batch_size,   input_size] | [batch_size,   input_size] | [batch_size,   input_size]
initial_hidden_state | - | [batch_size,   hidden_size] | [batch_size, hidden_size]
initial_cell_state | - | - | [batch_size,   hidden_size]
W | [1 * hidden_size,   input_size] | [3 *   hidden_size, input_size] | [4 *   hidden_size, input_size]
R | [1 * hidden_size,   hidden_size] | [3 *   hidden_size, hidden_size] | [4 *   hidden_size, hidden_size]
B | [1 * hidden_size] | [3 * hidden_size] (default) or [4 * hidden_size] | [4 *   hidden_size]
Outputs |   |   |  
Ho | [batch_size,   hidden_size] | [batch_size,   hidden_size] | [batch_size,   hidden_size]
Co | - | - | [batch_size,   hidden_size]

</div>

</body>

</html>

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=pl>&nbsp;</p>

<div style='direction:ltr'>

Inputs | RNN Seq | GRU Seq | LSTM Seq
-- | -- | -- | --
X | [batch_size,   seq_length, input_size] | [batch_size,   seq_length, input_size] | [batch_size,   seq_length, input_size]
initial_hidden_state (H) | [batch_size, num_directions, hidden_size] | [batch_size, num_directions, hidden_size] | [batch_size, num_directions, hidden_size]
initial_cell_state | - | - | [batch_size, num_directions, hidden_size]
sequence_lengths | [batch_size] | [batch_size] | [batch_size]
W | [num_directions, 1 * hidden_size,   input_size] | [num_directions, 3 *   hidden_size, input_size] | [num_directions, 4 *   hidden_size, input_size]
R | [num_directions, 1 * hidden_size,   hidden_size] | [num_directions, 3 *   hidden_size, hidden_size] | [num_directions, 4 *   hidden_size, hidden_size]
B | [num_directions, 1 * hidden_size] | [num_directions, 3 *   hidden_size] (default) or [num_directions, 4 * hidden_size] | [num_directions, 4 *   hidden_size]
Outputs |   |   |  
Y | [batch_size, num_directions, seq_len, hidden_size] | [batch_size, num_directions, seq_len, hidden_size] | [batch_size, num_directions, seq_len, hidden_size]
Ho | [batch_size, num_directions, hidden_size] | [batch_size, num_directions, hidden_size] | [batch_size, num_directions, hidden_size]
Co | - | - | [batch_size, num_directions, hidden_size]

</div>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=pl>&nbsp;</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=pl>&nbsp;</p>

</body>

</html>

### CODE REVIEW

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp [hunk 1]
```diff
@@ -692,6 +695,9 @@ const IStaticShapeInferFactory::TRegistry IStaticShapeInferFactory::registry{
     _OV_OP_SHAPE_INFER_VA_REG(ReduceMin, ShapeInferTA, op::util::ArithmeticReductionKeepDims, util::bit::mask(1)),
     _OV_OP_SHAPE_INFER_VA_REG(ReduceProd, ShapeInferTA, op::util::ArithmeticReductionKeepDims, util::bit::mask(1)),
     _OV_OP_SHAPE_INFER_VA_REG(ReduceSum, ShapeInferTA, op::util::ArithmeticReductionKeepDims, util::bit::mask(1)),
+    _OV_OP_SHAPE_INFER_MASK_REG(RNNCell, ShapeInferTA, util::bit::mask()),
+    _OV_OP_SHAPE_INFER_MASK_REG(RNNSequence, ShapeInferTA, util::bit::mask()),
+    _OV_OP_SHAPE_INFER_MASK_REG(Tile, ShapeInferenceTA, util::bit::mask(1)),
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp:700]: Added specific opset for each operation, but the default was working as well.

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+    // Merge hidden_size dimension across all inputs to evaluate output dimension
+    // `hidden_size` attribute is not used for backward compatibility
+    DimType merged_hidden_size = ht_pshape.rank().is_static() ? ht_pshape[1] : DimType();
+    for (size_t i = 2; i <= num_state_nodes; ++i) {
+        NODE_VALIDATION_CHECK(op,
+                              DimType::merge(merged_hidden_size,
+                                             merged_hidden_size,
+                                             input_shapes[i].rank().is_static() ? input_shapes[i][1] : DimType()),
+                              "Dimension `hidden_size` is not matched between inputs.");
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Ok, refactored.

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                    DimType::merge(merged_hidden_size, merged_hidden_size, input_shapes[i][0] / num_gates),
+                    "Dimension `hidden_size` is not matched between inputs.");
+            }
+        }
+    }
+
+    for (size_t i = 0; i < num_state_nodes; ++i) {  // Ho, Co outputs
+        output_shapes.push_back(TShape{merged_batch_size, merged_hidden_size});
+    }
+    return output_shapes;
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Done

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+std::vector<TShape> rnn_cell_base_shape_infer(const op::util::RNNCellBase* op,
+                                              const std::vector<TShape>& input_shapes,
+                                              size_t num_gates,
+                                              size_t num_state_nodes,
+                                              bool linear_before_reset = false) {
+    const auto num_inputs = 4 + num_state_nodes;
+    NODE_VALIDATION_CHECK(op, input_shapes.size() >= num_inputs, "Incorrect number of shapes has been provided.");
+
+    std::vector<TShape> output_shapes;
+    output_shapes.reserve(1 + num_state_nodes);
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Right, +1 is redundant here, updated to `output_shapes.reserve(num_state_nodes)`

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    }
+
+    // Y output
+    output_shapes.push_back(TShape{merged_batch_size,
+                                   merged_num_directions,
+                                   x_pshape.rank().is_static() ? x_pshape[1] : DimType(),
+                                   merged_hidden_size});
+    for (size_t i = 0; i < num_state_nodes; ++i) {  // Ho, Co outputs
+        output_shapes.push_back(TShape{merged_batch_size, merged_num_directions, merged_hidden_size});
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Done

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    } else if (m_direction == op::RecurrentSequenceDirection::BIDIRECTIONAL) {
+        valid_num_directions = 2;
+    } else {
+        NODE_VALIDATION_CHECK(op, false, "Attribute direction must be FORWARD or REVERSE or BIDIRECTIONAL.");
+    }
+
+    bool is_num_dir_valid = true;
+    DimType merged_num_directions = DimType(valid_num_directions);
+    for (size_t i = 1; i <= num_state_nodes; ++i) {
+        is_num_dir_valid = DimType::merge(merged_num_directions,
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Updated to `&=`, test will be added

File: src/core/shape_inference/include/rnn_base_shape_inference.hpp [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                                          merged_num_directions,
+                                          input_shapes[i].rank().is_static() ? input_shapes[i][1] : DimType());
+    }
+
+    for (size_t i = 2 + num_state_nodes; i < num_inputs; ++i) {
+        is_num_dir_valid = DimType::merge(merged_num_directions,
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/rnn_base_shape_inference.hpp]: Updated to `&=`, test will be added

File: src/core/shape_inference/include/lstm_sequence_shape_inference.hpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+#include "rnn_base_shape_inference.hpp"
+
+namespace ov {
+namespace op {
+namespace v5 {
+template <class TShape>
+std::vector<TShape> shape_infer(const LSTMSequence* op, const std::vector<TShape>& input_shapes) {
+    constexpr auto num_gates = 4;
+    constexpr auto num_state_nodes = 2;
+    return rnn::rnn_seq_base_shape_infer(op, input_shapes, num_gates, num_state_nodes, op->get_direction());
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/lstm_sequence_shape_inference.hpp]: Changed to `rnn::cell_base_shape_infer` and `rnn::seq_base_shape_infer`

File: src/core/tests/type_prop/rnn_seq_base.cpp [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+        }
+        op->set_arguments(inputs);
+        op->validate_and_infer_types();
+        return op;
+    }
+    return std::make_shared<T>(X, H_t, C_t, sequence_lengths, W, R, B, p.hidden_size.get_max_length(), p.direction);
+}
+
+template <class T>
+class RNNSeqBaseTest : public testing::Test {};
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/tests/type_prop/rnn_seq_base.cpp]: Done

File: src/plugins/intel_cpu/tests/unit/shape_inference_test/rnn_seq_shape_inference_test.cpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+//
+
+#include <gtest/gtest.h>
+
+#include "utils.hpp"
+
+using namespace ov;
+using namespace ov::intel_cpu;
+
+TEST(StaticShapeInferenceTest, RNNSequenceTest_default_ctor) {
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/plugins/intel_cpu/tests/unit/shape_inference_test/rnn_seq_shape_inference_test.cpp]: For me the simple tests are easier to read and faster to write without those fixtures,
but I refactored all of them to use the common OpStaticShapeInferenceTest as suggested.

**Review by praasz — APPROVED** [review]

**Review by iefode — APPROVED** [review]
LGTM from IE tests perspective

**Pull request merged**